### PR TITLE
[CustomerData Refactor] CustomerData has Payments and MeterReadings as EntityAdapters

### PR DESCRIPTION
**Pull request by julianrkung**
[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

# What's new in this PR
This PR changes `customerData` so that it's structurally identical to `inventoryData`. This is mostly done so that all slices in the redux store are stored similarly, and so operations done on different slices (such as adding, removing, etc...) look similar logically. The schema was also regenerated because 2 custom objects `meterReadings` and `payments` are sent from the backend.

This refactor was done now because we're about to complete functionality for payments and meter readings

Changelist
- `meterReadings` and `payments` are no longer stored inside of `customer` objects
- `meterReadings` and `payments` have their own EntityAdapters inside of `customerData`
- Schema was regenerated to fix foreign key issues
- Patches were regenerated to include new customer objects `meterReadings` and `payments`
- New selectors were created for `meterReadings` and `payments`
- Files `Home.tsx`, `CustomerMain.tsx`, `CustomerProfile.tsx` and `CustomerUtils.tsx` were refactored so that they used the selectors created for `meterReadings` and `payments` 

[//]: # "Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."

## How to review
1. Use the correct backend branch `julian/customer-data-refactor` PR'd [here](https://github.com/calblueprint/meepanyar-node/pull/22)
2. Open the app and clear site data to reset redux
3. Log in and make sure there are no errors on the Home, CustomerMain, CustomerProfile, or Records screens
4. Create a customer and make sure that they are still properly made on the frontend and backend
5. Edit a customer and make sure they're still properly editable
6. Go to the customer's records and make sure that meter readings and payments are populated

[//]: # 'The order in which to review files and what to expect when testing locally'

### Online sources

[//]: # 'Optional - copy links to any tutorial or documentation that was useful to you when working on this PR'

### Related PRs

[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

## Next steps
- Refactor sites
- Functionality for payments and meter readings

[//]: # "What's NOT in this PR, doesn't work yet, and/or still needs to be done"

### Screenshots

[//]: # "Add screenshots of expected behavior - GIFs if you're feeling fancy!"

CC: @julianrkung

[//]: # 'This tags Julian as a default. Feel free to change, or add on anyone who you should be in on the conversation.'

### CODE REVIEW

File: src/lib/redux/customerData.ts [hunk 1]
```diff
@@ -21,9 +36,76 @@ export const setCurrentCustomerIdInRedux = (customerId: string): void => {
...
+    selectAllCustomersArray,
+    (allMeterReadingsInCurrentPeriod, allCustomers) => {
+        const meteredCustomerIds = new Set(allMeterReadingsInCurrentPeriod.map(meterReading => meterReading.customerId));
+        return allCustomers.filter(customer => !meteredCustomerIds.has(customer.id))
+    }
+)
+
+// For each customer who has been metered this period, 
+// we calculate the amount
+// the amount they need to pay
```
Comment (wangannie) [inline src/lib/redux/customerData.ts]: Nit: some funky wording here 👀

File: src/lib/redux/customerDataSlice.ts [hunk 2]
```diff
@@ -14,31 +63,34 @@ export const EMPTY_CUSTOMER: CustomerRecord = {
...
+    meterNumber: -1,
+    customerId: '',
+}
 
-// TODO: @julianrkung, We should eventually make this a Record<SiteId, Record<CustomerId, CustomerRecord>> map for efficiency
-// We can't do this yet because customers created while offline do not receive an id yet
-interface customerDataSliceState {
-    sitesCustomers: Record<SiteId, EntityState<CustomerRecord>>;
-    currentCustomerId: string;
+export const EMPTY_CUSTOMER_DATA: CustomerData = {
```
Comment (wangannie) [inline src/lib/redux/customerDataSlice.ts]: This could be confusing with the existing `EMPTY_CUSTOMER`, so it may make more sense to rename this to `EMPTY_SITE_CUSTOMER_DATA` (at least I had it called `EMPTY_SITE_INVENTORY_DATA` in inventory) to emphasize that represents all customer data for a site.

File: src/lib/redux/customerDataSlice.ts [hunk 3]
```diff
@@ -1,9 +1,58 @@
...
+
+export const {
+    selectEntities: selectAllMeterReadings,
+    selectAll: selectAllMeterReadingsArray,
+    selectById: selectMeterReadingById,
+    selectIds: selectMeterReadingIds
+} = meterReadingsAdapter.getSelectors(
+    (state: RootState) => state.customerData.sitesCustomers[state.siteData.currentSite.id].meterReadings);
+
+interface CustomerData {
```
Comment (wangannie) [inline src/lib/redux/customerDataSlice.ts]: Maybe rename to more explicitly mention that this is customer data related to a specific site. I called it `SiteInventoryData` in inventory to delineate the difference.

File: src/lib/utils/customerUtils.ts [hunk 4]
```diff
@@ -1,50 +1,33 @@
...
-const isCurrentReading = (mr: MeterReadingRecord): boolean => {
-    const period: string = getPeriodFromDate(mr.date);
-    const currPer: string = getCurrentPeriod();
-    return period === currPer;
-};
+import { isBeforeCurrentPeriod } from '../../lib/moment/momentUtils';
+import { selectMeterReadingsByCustomerId } from '../../lib/redux/customerData';
+import { store } from '../redux/store';
 
 // Returns last reading in the period
```
Comment (wangannie) [inline src/lib/utils/customerUtils.ts]: This isn't related to period anymore right? Probably good to update the comment

File: src/lib/utils/customerUtils.ts [hunk 5]
```diff
@@ -1,50 +1,33 @@
...
 // Could be undefined if no other reading exists except current
 export const getStartingReading = (customer: CustomerRecord): MeterReadingRecord | undefined => {
-  let mostRecentReading: MeterReadingRecord | undefined;
-  for (let i = 0; i < customer.meterReadings.length; i += 1) {
-    const thisReading = customer.meterReadings[i];
-    if (!isCurrentReading(thisReading)) {
-      mostRecentReading = getClosestReading(thisReading, mostRecentReading);
+  const meterReadings: MeterReadingRecord[] = selectMeterReadingsByCustomerId(store.getState(), customer.id);
+
+  // Meter readings are
```
Comment (wangannie) [inline src/lib/utils/customerUtils.ts]: are..

File: src/lib/redux/customerData.ts [hunk 6]
```diff
@@ -21,9 +36,76 @@ export const setCurrentCustomerIdInRedux = (customerId: string): void => {
...
+                    .filter(meterReading => meterReading.customerId === customer.id)
+                    .reduce((totalAmountOwed, meterReading: MeterReadingRecord) => totalAmountOwed + meterReading.amountBilled, 0);
+
+                const amountPaidThisPeriod = periodPayments
+                    .filter(payments => payments.billedToId === customer.id)
+                    .reduce((totalAmountPaid, payment: PaymentRecord) => totalAmountPaid + payment.amount, 0);
+
+                if (amountOwedThisPeriod <= amountPaidThisPeriod) {
+                    customersToCollect.push(customer)
+                }
```
Comment (wangannie) [inline src/lib/redux/customerData.ts]: Since this is such a fat selector, I'm wondering if you could extract some parts of this out to helper functions, like maybe what's happening in this forEach? Although that may end up negating the performance benefits of selectors.

**Review by wangannie — APPROVED** [review]
This is looking really good! I just noted a few suggestions on commenting + renaming for clarity but I didn't run into any issues testing.

File: src/lib/redux/customerData.ts (hunk 6 above)
Comment (julianrkung) [reply src/lib/redux/customerData.ts]: Good idea, I was afraid that this function might be hard to follow. Created selectors to calculate `amountOwedThisPeriod` and `amountPaidThisPeriod`.

File: src/lib/redux/customerDataSlice.ts (hunk 3 above)
Comment (julianrkung) [reply src/lib/redux/customerDataSlice.ts]: Renamed to `SiteCustomerData` to match, thanks!
